src/main.rs

```diff
@@ -1,100 +1,119 @@
-            // Generate private keys, reporting for each match against vanity prefix.
-            loop {
-                let private_key = key_generator.generate_ed25519_private_key();
-                // Assume account address is identical to authentication key (no rotation).
-                let account_address_bytes = auth_key_bytes_vec(&private_key);
-                // Get bytes to check against: if multisig, derive multisig account address.
-                let search_bytes = if multisig {
-                    // Note that for performance this moves (rather than clones) address bytes, such
-                    // that account address must be re-derived from private key if a match is found.
-                    create_multisig_account_address(
-                        AccountAddress::from_bytes(account_address_bytes).unwrap(),
-                        SEQUENCE_NUMBER,
-                    )
-                    .to_vec()
-                } else {
-                    account_address_bytes // If standard account, check against account address.
-                };
-                // If a match found, print diagnostics then keep searching.
-                if search_bytes.starts_with(&prefix_bytes) {
-                    // Get search bytes as a string.
-                    let search_str = hex::encode(search_bytes);
-                    // Verify full string match if prefix is an odd number of characters.
-                    if is_odd_length && !search_str.starts_with(&prefix_str) {
-                        continue;
-                    }
-                    // Get standard account address. If multisig:
-                    let account_address_hex = if multisig {
-                        println!("Multisig account address: 0x{}", search_str);
-                        // Re-derive account address bytes here rather than cloning during search.
-                        hex::encode(auth_key_bytes_vec(&private_key))
-                    } else {
-                        search_str // If not multisig, standard account address is from search.
-                    };
-                    println!("Standard account address: 0x{}", account_address_hex);
-                    println!(
-                        "Private key:              0x{}\n",
-                        hex::encode(private_key.to_bytes())
-                    );
-                }
-            }
-        });
+
+    if args.prefix.len() % 2 == 1 {
+        hex::decode(format!("{}0", &args.prefix))?;
+    } else {
+        hex::decode(&args.prefix)?;
     }
-    // Run infinite loop to prevent main thread from returning and closing search threads.
+
+    Ok(args)
+}
+
+fn generate_key(
+    prefix: String,
+    multisig: bool,
+    res_tx: channel::Sender<(String, String)>,
+    exit_rx: channel::Receiver<()>,
+) -> Result<()> {
+    let mut key_generator = KeyGen::from_os_rng();
+    let prefix_bytes = hex::decode(&prefix)?;
```

Comment:
This line will fail if the prefix is an odd number of characters, which is a valid input

The original algorithm was designed in two phases:

1. Bytewise search against even number of characters
2. If a match found against bytes, then an additional string compare

This accelerates the search while minimizing string operations